### PR TITLE
[XLA:CPU][oneDNN] Update oneDNN LayerNorm and Softmax

### PR DESCRIPTION
**Pull request by sachinmuradi**
This PR addresses following : 
1) Add support for FP16 to oneDNN softmax pattern matching
2) Adds another pattern for oneDNN layernorm based on keras layernorm operation
3) Fixes previous oneDNN layernorm pattern, as it was focused on flax layernorm pattern, which recently wen through an update while moment calculation
4) Adds FP16 support to oneDNN layernorm pattern matching.

### CODE REVIEW

File: xla/service/cpu/onednn_ops_rewriter.cc [hunk 1]
```diff
@@ -143,183 +139,288 @@ std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
...
+  // variance = Mean((X - Mean(x))^2)
+  // Z = scale / sqrt(variance + eps)
+  // LN(X) = X*Z + Bias - Mean(X)*Z
+
+  HloInstruction *src_a, *src_b, *src_c;
+  HloInstruction *bias_node, *scaled_norm_a, *scaled_norm_b, *mean0_a, *epsilon,
+      *sqrd_diff_mean, *scale_node, *sqrd_diff;
+
+  // First Match X*Z + Bias - Mean(X)*Z
+
```
Comment (penpornk) [inline xla/service/cpu/onednn_ops_rewriter.cc:177]: Nit: Please remove this empty line.

File: xla/service/cpu/onednn_ops_rewriter.cc [hunk 2]
```diff
@@ -143,183 +139,288 @@ std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
...
+                  .WithOneUser()))) {
+    return std::nullopt;
+  }
+
+  if (scaled_norm_a != scaled_norm_b) return std::nullopt;
+
+  const Shape& src_shape = (*src)->shape();
+  if (!IsSupportedType(src_shape.element_type())) return std::nullopt;
+
+  // get bias
```
Comment (penpornk) [inline xla/service/cpu/onednn_ops_rewriter.cc:197]: Nit: Could you please capitalize the first letter of each comment?
```suggestion
  // Get bias
```

File: xla/service/cpu/onednn_ops_rewriter.cc [hunk 3]
```diff
@@ -143,183 +139,288 @@ std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
...
+
+  HloInstruction* shift = FindLayerNormShift(instr);
+  shiftFound = (shift != nullptr);
+
+  HloInstruction* scale_gamma = FindLayerNormScale(hinge);
+  scaleFound = (scale_gamma != nullptr);
+
+  // Currently patterns without scale and shift are
+  // not supported.
+  // OneDNN only supports 2 <= rank <= 5
```
Comment (penpornk) [inline xla/service/cpu/onednn_ops_rewriter.cc:280]: Nit: We can fit the first two lines in one line.
```suggestion
  // Currently patterns without scale and shift are not supported.
  // OneDNN only supports 2 <= rank <= 5
```

File: xla/service/cpu/onednn_ops_rewriter.cc [hunk 4]
```diff
@@ -143,183 +139,288 @@ std::optional<HloInstruction*> MatchSoftmax(HloInstruction* instr) {
...
+    HloInstruction *src, *scale, *bias;
+    float eps;
+    bool found_ln = false;
+    bool is_bf16orfp16_convert = false;
+    bool is_producer_bf16orfp16 = false;
+    HloInstruction* convert_instr;
+
+    if (MatchTFKerasLayerNorm(instr, &src, &scale, &bias, &eps)) {
+      found_ln = true;
     }
```
Comment (penpornk) [inline xla/service/cpu/onednn_ops_rewriter.cc:398]: Nit: This can be simplified to
```suggestion
    bool found_ln = MatchTFKerasLayerNorm(instr, &src, &scale, &bias, &eps));
```

File: xla/tests/onednn_layer_norm_test.cc [hunk 5]
```diff
@@ -92,66 +92,134 @@ TEST_F(LayerNormTest, SimpleTest) {
   )");
 }
 
-TEST_F(LayerNormTest, SimpleTestBF16) {
+TEST_F(LayerNormTest, LayerNormTest0_BF16) {
   const char* layer_norm_module_str = R"(
-  HloModule layer_norm_bf16.test, entry_computation_layout={(f32[768]{0}, f32[768]{0}, bf16[16,128,768]{2,1,0})->bf16[16,128,768]{2,1,0}}, allow_spmd_sharding_propagation_to_output={true}
+  HloModule layer_norm.test, entry_computation_layout={(bf16[84,197,768]{2,1,0}, f32[768]{0}, f32[768]{0})->bf16[84,197,768]{2,1,0}}
```
Comment (penpornk) [inline xla/tests/onednn_layer_norm_test.cc:97]: Since the module strings for BF16 and FP16 are just the FP32 one with minor modifications, can we have just one base string as a template for all three tests (FP32, BF16, and FP16) and just do some string manipulations to change the dtype / add conversion?

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for the PR!

File: xla/service/cpu/backend_config.proto [hunk 6]
```diff
@@ -44,5 +45,5 @@ message OneDnnLayerNormConfig {
     SHIFT = 2;
     SCALE_AND_SHIFT = 3;
   }
-  FusionKind fused_ops = 1;
+  FusionKind fused_ops = 2;
```
[the remark this answers is not in the document]
Comment (sachinmuradi) [reply xla/service/cpu/backend_config.proto]: Done

File: xla/service/cpu/onednn_ops_rewriter.cc (hunk 2 above)
Comment (sachinmuradi) [reply xla/service/cpu/onednn_ops_rewriter.cc]: Done

File: xla/service/cpu/onednn_ops_rewriter.cc (hunk 4 above)
Comment (sachinmuradi) [reply xla/service/cpu/onednn_ops_rewriter.cc]: Done

File: xla/service/cpu/onednn_ops_rewriter.cc (hunk 1 above)
Comment (sachinmuradi) [reply xla/service/cpu/onednn_ops_rewriter.cc]: Done

File: xla/tests/onednn_layer_norm_test.cc (hunk 5 above)
Comment (sachinmuradi) [reply xla/tests/onednn_layer_norm_test.cc]: I modified the test and tried to address this by creating some common strings for HLO texts that can be used on all dtype tests, let me know if that looks fine.

**Review by penpornk — APPROVED** [review]
Thank you for the changes and sorry for the delay!
